src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -0,0 +1,177 @@
+
+  // Attributes interact with "canActAs" because if "Y canActAs X"
+  // then Y also picks up X's attributes. We need to generate
+  // an additional rule to implement this behavior. If the attribute
+  // under translation is `X PredX`, the additional rule is:
+  // `speaker says Y PredX :-
+  //    speaker says Y canActAs X, speaker says X PredX`
+  // (Where Y is a fresh variable)
+  Principal prin_y = Principal(FreshVar());
+
+  // This is `speaker says Y PredX`
+  Predicate generated_lhs = PushPrincipal("says_", prin_y, main_predicate);
+
+  Predicate y_can_act_as_x =
+      Predicate("canActAs", {prin_y.name(), attribute.principal().name()},
+                Sign::kPositive);
+
+  Predicate speaker_says_y_can_act_as_x =
+      PushPrincipal("says_", speaker, y_can_act_as_x);
+
+  // This is `speaker says X PredX`
+  Predicate speaker_says_x_pred =
+      PushPrincipal("says_", speaker, main_predicate);
+
+  // This is the full generated rule:
+  // `speaker says Y PredX :-
+  //    speaker says Y canActAs X, speaker says X PredX`
+  return DLIRAssertion(
+      DLIRCondAssertion(generated_lhs, {std::move(speaker_says_y_can_act_as_x),
+                                        std::move(speaker_says_x_pred)}));
+}
+
+DLIRAssertion LoweringToDatalogPass::SpokenCanActAsToDLIR(
+    const Principal& speaker, const CanActAs& can_act_as) {
+  Predicate main_predicate = CanActAsToDLIR(can_act_as);
+
+  // "canActAs" facts are passed to principals via other canActAs facts in
+  // essentially the same way as attributes. This function adds extra
+  // rules to pass these around. If the `canActAs` under translation
+  // is `X canActAs Z`, then the rule we need to generate is:
+  // `speaker says Y PredX :-
+  //    speaker says Y canActAs X, speaker says X canActAsZ`
+  // (Where Y is a fresh variable)
+  Principal prin_y = Principal(FreshVar());
+
+  // This is `speaker says Y PredX`
+  Predicate generated_lhs = PushPrincipal("says_", prin_y, main_predicate);
+
+  Predicate y_can_act_as_x =
+      Predicate("canActAs", {prin_y.name(), can_act_as.left_principal().name()},
+                Sign::kPositive);
+
+  Predicate speaker_says_y_can_act_as_x =
+      PushPrincipal("says_", speaker, y_can_act_as_x);
+
+  // This is `speaker says X canActAs Z`
+  Predicate speaker_says_x_can_act_as_z =
+      PushPrincipal("says_", speaker, main_predicate);
+
+  // This is the full generated rule:
+  // `speaker says Y PredX :-
+  //    speaker says Y canActAs X, speaker says X canActAs Z`
+  return DLIRAssertion(DLIRCondAssertion(
+      generated_lhs, {std::move(speaker_says_y_can_act_as_x),
+                      std::move(speaker_says_x_can_act_as_z)}));
+}
+
+std::pair<Predicate, std::vector<DLIRAssertion>>
+LoweringToDatalogPass::FactToDLIR(const Principal& speaker, const Fact& fact) {
+  return std::visit(
+      raksha::utils::overloaded{
+          [this, &speaker](BaseFact base_fact) {
+            return BaseFactToDLIR(speaker, base_fact);
```

Comment:
Ideally this should be a function in `BaseFact`. OK to defer to later as it is in the same spirit of exposing `std::variant`.

Reply:
I'm assuming [this issue](https://github.com/google-research/raksha/issues/345) has resolved this for now, so I am resolving this. (?)